![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)


# Proyecto: Riesgo en el Banco Giturra

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### Cuerpo Docente:

- Profesor: Pablo Badilla, Ignacio Meza De La Jara
- Auxiliar: Sebastián Tinoco
- Ayudante: Diego Cortez M., Felipe Arias T.

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

## Reglas

- Fecha de entrega: 01/06/2021
- **Grupos de 2 personas.**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Estrictamente prohibida la copia.
- Pueden usar cualquier material del curso que estimen conveniente.


---


# Presentación del Problema


![](https://www.diarioeldia.cl/u/fotografias/fotosnoticias/2019/11/8/67218.jpg)


**Giturra**, un banquero astuto y ambicioso, estableció su propio banco con el objetivo de obtener enormes ganancias. Sin embargo, su reputación se vio empañada debido a las tasas de interés usureras que imponía a sus clientes. A medida que su banco crecía, Giturra enfrentaba una creciente cantidad de préstamos impagados, lo que amenazaba su negocio y su prestigio.

Para abordar este desafío, Giturra reconoció la necesidad de reducir los riesgos de préstamo y mejorar la calidad de los préstamos otorgados. Decidió aprovechar la ciencia de datos y el análisis de riesgo crediticio. Contrató a un equipo de expertos para desarrollar un modelo predictivo de riesgo crediticio.

Cabe señalar que lo modelos solicitados por el banquero deben ser interpretables. Ya que estos le permitira al equipo comprender y explicar cómo se toman las decisiones crediticias. Utilizando visualizaciones claras y explicaciones detalladas, pudieron identificar las características más relevantes, le permitirá analizar la distribución de la importancia de las variables y evaluar si los modelos son coherentes con el negocio.

Para esto Giturra les solicita crear un modelo de riesgo disponibilizandoles una amplia gama de variables de sus usuarios: como historiales de crédito, ingresos y otros factores financieros relevantes, para evaluar la probabilidad de incumplimiento de pago de los clientes. Con esta información, Giturra podra tomar decisiones más informadas en cuanto a los préstamos, ofreciendo condiciones más favorables a aquellos con menor riesgo de impago.


## Instalación de Librerías y Carga de Datos.


Para el desarrollo de su proyecto, utilice el conjunto de datos `dataset.pq` para entrenar un modelo de su elección. Además, se adjunta junto con los datos del proyecto un archivo llamado `requirements.txt` que contiene todas las bibliotecas y versiones necesarias para el desarrollo del proyecto. Se le recomienda levantar un ambiente de `conda` para instalar estas librerías y así evitar cualquier problema con las versiones.


In [497]:
!  pip install -r requirements.txt

In [498]:
# Data Handle
import pandas as pd
import numpy as np
import random

# Viz
#import matplotlib.pyplot as plt
#import seaborn as sns
#import plotly.express as px

# Seed
seed = 42

import warnings
warnings.filterwarnings("ignore")

In [499]:
! pip install pyarrow
! pip install fastparquet
! pip install plotly

## 1 Introducción

## 2 Carga de Datos y EDA

In [500]:
df = pd.read_parquet('dataset.pq')
df

customer_id   age     occupation  annual_income  monthly_inhand_salary  \
0       CUS_0xd40  23.0      Scientist       19114.12            1824.843333   
1      CUS_0x21b1  28.0        Teacher       34847.84            3037.986667   
2      CUS_0x2dbc  34.0       Engineer      143162.64           12187.220000   
3      CUS_0xb891  55.0   Entrepreneur       30689.89            2612.490833   
4      CUS_0x1cdb  21.0      Developer       35547.71            2853.309167   
...           ...   ...            ...            ...                    ...   
12495  CUS_0x372c  19.0         Lawyer       42903.79            3468.315833   
12496   CUS_0xf16  45.0  Media_Manager       16680.35                    NaN   
12497  CUS_0xaf61  50.0         Writer       37188.10            3097.008333   
12498  CUS_0x8600  29.0      Architect       20002.88            1929.906667   
12499  CUS_0x942c  25.0       Mechanic       39628.99            3359.415833   

       num_bank_accounts  num_credit_card  interest_rate  num_of_loan  \
0                      3                4              3          4.0   
1                      2                4              6          1.0   
2                      1                5              8          3.0   
3                      2                5              4       -100.0   
4                      7                5              5       -100.0   
...                  ...              ...            ...          ...   
12495                  0                4              6          1.0   
12496                  1                1              5          4.0   
12497                  1                4              5          3.0   
12498                 10                8             29          5.0   
12499                  4                6              7          2.0   

       delay_from_due_date  ...  num_credit_inquiries  outstanding_debt  \
0                        3  ...                   4.0            809.98   
1                        3  ...                   2.0            605.03   
2                        8  ...                   3.0           1303.01   
3                        4  ...                   4.0            632.46   
4                        1  ...                   4.0            943.86   
...                    ...  ...                   ...               ...   
12495                    9  ...                   1.0           1079.48   
12496                    1  ...                   8.0            897.16   
12497                    7  ...                   3.0            620.64   
12498                   33  ...                   9.0           3571.70   
12499                   18  ...                   3.0            502.38   

       credit_utilization_ratio  credit_history_age  payment_of_min_amount  \
0                     23.933795                 NaN                     No   
1                     32.933856                27.0                     No   
2                     38.374753                18.0                     No   
3                     27.332515                17.0                     No   
4                     25.862922                31.0                    Yes   
...                         ...                 ...                    ...   
12495                 35.716618                28.0                     No   
12496                 41.212367                 NaN                     No   
12497                 39.300980                30.0                     No   
12498                 37.140784                 6.0                    Yes   
12499                 34.192463                31.0                     No   

       total_emi_per_month amount_invested_monthly  \
0                49.574949               24.785217   
1                18.816215              218.904344   
2               246.992319            10000.000000   
3                16.415452              125.617251   
4                 0.000000              181.330901   
...   

In [501]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12500 entries, 0 to 12499
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   customer_id               12500 non-null  object 
 1   age                       12500 non-null  float64
 2   occupation                12500 non-null  object 
 3   annual_income             12500 non-null  float64
 4   monthly_inhand_salary     10584 non-null  float64
 5   num_bank_accounts         12500 non-null  int64  
 6   num_credit_card           12500 non-null  int64  
 7   interest_rate             12500 non-null  int64  
 8   num_of_loan               12500 non-null  float64
 9   delay_from_due_date       12500 non-null  int64  
 10  num_of_delayed_payment    11660 non-null  float64
 11  changed_credit_limit      12246 non-null  float64
 12  num_credit_inquiries      12243 non-null  float64
 13  outstanding_debt          12500 non-null  float64
 14  credit

In [502]:
df.describe().T.round(1)

count          mean           std           min  \
age                       12500.0  1.058000e+02  6.645000e+02 -5.000000e+02   
annual_income             12500.0  1.616206e+05  1.297842e+06  7.005900e+03   
monthly_inhand_salary     10584.0  4.186600e+03  3.173700e+03  3.036000e+02   
num_bank_accounts         12500.0  1.690000e+01  1.144000e+02 -1.000000e+00   
num_credit_card           12500.0  2.320000e+01  1.320000e+02  0.000000e+00   
interest_rate             12500.0  7.320000e+01  4.687000e+02  1.000000e+00   
num_of_loan               12500.0  3.100000e+00  6.510000e+01 -1.000000e+02   
delay_from_due_date       12500.0  2.110000e+01  1.490000e+01 -5.000000e+00   
num_of_delayed_payment    11660.0  3.290000e+01  2.374000e+02 -3.000000e+00   
changed_credit_limit      12246.0  1.040000e+01  6.800000e+00 -6.500000e+00   
num_credit_inquiries      12243.0  2.630000e+01  1.818000e+02  0.000000e+00   
outstanding_debt          12500.0  1.426200e+03  1.155200e+03  2.000000e-01   
credit_utilization_ratio  12500.0  3.230000e+01  5.200000e+00  2.010000e+01   
credit_history_age        11380.0  1.820000e+01  8.300000e+00  0.000000e+00   
total_emi_per_month       12500.0  1.488400e+03  8.561400e+03  0.000000e+00   
amount_invested_monthly   11914.0  6.388000e+02  2.049200e+03  0.000000e+00   
monthly_balance           12145.0 -2.744614e+22  3.024684e+24 -3.333333e+26   
credit_score              12500.0  3.000000e-01  5.000000e-01  0.000000e+00   

                              25%      50%      75%         max  
age                          25.0     33.0     42.0      8678.0  
annual_income             19453.3  37572.4  72690.2  23834698.0  
monthly_inhand_salary      1622.4   3087.6   5967.9     15204.6  
num_bank_accounts             3.0      6.0      7.0      1756.0  
num_credit_card               4.0      5.0      7.0      1499.0  
interest_rate                 8.0     14.0     20.0      5789.0  
num_of_loan                   1.0      3.0      5.0      1495.0  
delay_from_due_date          10.0     18.0     28.0        67.0  
num_of_delayed_payment        9.0     14.0     18.0      4293.0  
changed_credit_limit          5.4      9.4     14.9        37.0  
num_credit_inquiries          4.0      6.0     10.0      2554.0  
outstanding_debt            566.1   1166.2   1946.0      4998.1  
credit_utilization_ratio     28.1     32.4     36.6        48.2  
credit_history_age           12.0     18.0     25.0        33.0  
total_emi_per_month          31.5     72.9    169.6     81971.0  
amount_invested_monthly      73.7    134.1    261.7     10000.0  
monthly_balance             270.2    339.4    471.4      1463.8  
credit_score                  0.0      0.0      1.0         1.0

Outliers en:
- age: valores negativos y extremos
- annual_income: valores extremos
- num_bank_accounts: valores negativos y extremos
- num_of_loan: valores negativos y extremos
- delay_from_due_date: valores negativos y extremos
- num_of_delayed_payment: valores negativos y extremos
- changed_credit_limit: valores negativos
- num_credit_inquiries: valores extremos
- total_emi_per_month:valores extremos
- monthly_balance:  extremos

In [503]:
mask = df.loc[:,'age'] <= 18
df.loc[mask,:]

customer_id   age    occupation  annual_income  monthly_inhand_salary  \
36     CUS_0x4080  16.0      Mechanic      29469.980            2227.831667   
45     CUS_0x8e9b  16.0  Entrepreneur      55829.790                    NaN   
51     CUS_0xb986  14.0     Developer      39887.220            3224.935000   
58     CUS_0x47db  16.0      Musician      78988.480            6449.373333   
65     CUS_0x5cdf  18.0      Musician      75273.240                    NaN   
...           ...   ...           ...            ...                    ...   
12344  CUS_0xbf23  17.0    Accountant       8246.785             582.232083   
12393  CUS_0x50d4  18.0      Engineer      41159.540            3638.961667   
12412  CUS_0x6c58  18.0    Accountant       9409.450             952.120833   
12438  CUS_0x504e  17.0       _______      72572.460            6233.705000   
12475  CUS_0x1232  17.0     Developer      14937.490            1142.790833   

       num_bank_accounts  num_credit_card  interest_rate  num_of_loan  \
36                     7                7             24          5.0   
45                    10               10             18          8.0   
51                     9                9             16       -100.0   
58                     9                5             29          7.0   
65                     8                6             32          5.0   
...                  ...              ...            ...          ...   
12344                 10                9             22          4.0   
12393                  9                6             16          7.0   
12412                  5                6             13          7.0   
12438                  4                4              7          4.0   
12475                  8                8             22          7.0   

       delay_from_due_date  ...  num_credit_inquiries  outstanding_debt  \
36                      53  ...                  11.0           3421.66   
45                      30  ...                  11.0           3422.49   
51                      57  ...                  14.0           3119.60   
58                      27  ...                   9.0           1746.90   
65                      50  ...                   8.0           2497.34   
...                    ...  ...                   ...               ...   
12344                   39  ...                  10.0           1528.01   
12393                   31  ...                  10.0           2370.64   
12412                    7  ...                   7.0            209.02   
12438                   29  ...                   8.0           1321.46   
12475                   43  ...                   9.0           3699.44   

       credit_utilization_ratio  credit_history_age  payment_of_min_amount  \
36                    32.962950                13.0                    Yes   
45                    37.865668                11.0                    Yes   
51                    34.970866                13.0                     NM   
58                    28.955438                10.0                    Yes   
65                    39.124418                 5.0                    Yes   
...                         ...                 ...                    ...   
12344                 32.856037                 9.0                    Yes   
12393                 22.415697                19.0                    Yes   
12412                 40.709515                 8.0                    Yes   
12438                 24.551789                20.0                     NM   
12475                 23.120963                13.0                     NM   

       total_emi_per_month amount_invested_monthly  \
36               69.685459               24.066131   
45              314.901785              255.729501   
51              133.470845              145.940549   
58              291.782760              162.081601   
65              351.367045              369.491720   
...               

In [504]:
mask = df.loc[:,'age'] >90
df.loc[mask,:]

customer_id     age    occupation  annual_income  monthly_inhand_salary  \
50      CUS_0xb14  3052.0       Manager       49967.01            4091.917500   
67     CUS_0xa156  4431.0  Entrepreneur       58674.66                    NaN   
147    CUS_0x710f  3115.0     Scientist       33502.10            2890.841667   
343    CUS_0x8c03  4670.0     Scientist       54987.52                    NaN   
373    CUS_0xa560   395.0       _______      133410.66           11387.555000   
...           ...     ...           ...            ...                    ...   
12083  CUS_0x6ff5  7445.0  Entrepreneur       43856.38            3870.698333   
12277  CUS_0x42c3   598.0  Entrepreneur       40056.10                    NaN   
12398  CUS_0x6334  1683.0      Mechanic    13944362.00            8216.153333   
12436  CUS_0x290c  1070.0      Engineer       28954.04            2703.836667   
12443  CUS_0x2637  5798.0        Writer       29090.50            2472.208333   

       num_bank_accounts  num_credit_card  interest_rate  num_of_loan  \
50                     6                5             31          6.0   
67                     8                5             15          4.0   
147                    7                9             27          6.0   
343                    4                5             10          0.0   
373                    1                3              4          0.0   
...                  ...              ...            ...          ...   
12083                  2                5              6          0.0   
12277                  6                4             13          3.0   
12398                  5                6             15          5.0   
12436                  8                5             25          5.0   
12443                  5                7           5616          4.0   

       delay_from_due_date  ...  num_credit_inquiries  outstanding_debt  \
50                      27  ...                  12.0           2253.95   
67                      55  ...                  11.0           2425.38   
147                     57  ...                  12.0           2362.68   
343                     18  ...                   8.0           1263.12   
373                      5  ...                   3.0            933.84   
...                    ...  ...                   ...               ...   
12083                    4  ...                   1.0            975.82   
12277                   29  ...                   2.0            182.58   
12398                   25  ...                   4.0            698.23   
12436                   36  ...                  10.0           2077.68   
12443                   11  ...                   6.0           1612.70   

       credit_utilization_ratio  credit_history_age  payment_of_min_amount  \
50                    31.520782                17.0                    Yes   
67                    33.808584                13.0                     NM   
147                   22.183579                18.0                    Yes   
343                   29.000913                16.0                    Yes   
373                   27.005230                27.0                     No   
...                         ...                 ...                    ...   
12083                 29.259389                16.0                     No   
12277                 29.849686                 NaN                     No   
12398                 24.567582                14.0                    Yes   
12436                 37.483994                16.0                    Yes   
12443                 39.668508                16.0                    Yes   

       total_emi_per_month amount_invested_monthly  \
50              156.003312              116.361294   
67              116.103417              334.269165   
147             158.607927              184.548187   
343               0.000000              420.219964   
373               0.000000              518.2357

In [505]:
(df.isna().sum()/len(df)).round(3)*100

customer_id                  0.0
age                          0.0
occupation                   0.0
annual_income                0.0
monthly_inhand_salary       15.3
num_bank_accounts            0.0
num_credit_card              0.0
interest_rate                0.0
num_of_loan                  0.0
delay_from_due_date          0.0
num_of_delayed_payment       6.7
changed_credit_limit         2.0
num_credit_inquiries         2.1
outstanding_debt             0.0
credit_utilization_ratio     0.0
credit_history_age           9.0
payment_of_min_amount        0.0
total_emi_per_month          0.0
amount_invested_monthly      4.7
payment_behaviour            0.0
monthly_balance              2.8
credit_score                 0.0
dtype: float64

nulos en:
- monthly_inhand_salary
- num_of_delayed_payment
- changed_credit_limit
- num_credit_inquiries
- credit_history_age
- amount_invested_monthly
- monthly_balance

La variable monthly_inhand_salary, num_of_delayed_payment y credit_history_age contienen más de un 5% de valores nulos, por lo que se descartarán, se considerará rellenar sólo aquellas que tengan menos de este porcentaje.

In [506]:
df.loc[:,'customer_id'].duplicated().sum()

0

In [507]:
df.loc[:,'occupation'].unique()

array(['Scientist', 'Teacher', 'Engineer', 'Entrepreneur', 'Developer',
       'Lawyer', 'Media_Manager', 'Doctor', 'Journalist', 'Manager',
       'Accountant', 'Musician', 'Mechanic', 'Writer', 'Architect',
       '_______'], dtype=object)

In [508]:
mask = df.loc[:,'occupation'] == '_______'
df.loc[mask,:]

customer_id   age occupation  annual_income  monthly_inhand_salary  \
37     CUS_0x706a  20.0    _______      72559.360            6284.613333   
44     CUS_0xaedb  20.0    _______      85554.030            7185.502500   
46     CUS_0x609d  27.0    _______      14165.230            1057.435833   
60     CUS_0x7d0b  22.0    _______      19981.600            1459.133333   
64     CUS_0xab76  26.0    _______      60162.100            5197.508333   
...           ...   ...        ...            ...                    ...   
12441  CUS_0x85dd  45.0    _______       8974.555             783.879583   
12456  CUS_0x2e91  31.0    _______      70926.060            5741.505000   
12469  CUS_0x4a8f  24.0    _______      34493.920            3043.493333   
12491  CUS_0xb11c  38.0    _______      15319.650            1460.637500   
12492  CUS_0xad4f  48.0    _______      22620.790                    NaN   

       num_bank_accounts  num_credit_card  interest_rate  num_of_loan  \
37                     4                5             17          4.0   
44                     4                2              3          0.0   
46                     7               10             33          9.0   
60                    10                8             20          7.0   
64                     5                7              5          7.0   
...                  ...              ...            ...          ...   
12441                 10                8             28          7.0   
12456                  7                7             10          2.0   
12469                  9               10             24          7.0   
12491                  6                7             15          4.0   
12492                  6                2              9          0.0   

       delay_from_due_date  ...  num_credit_inquiries  outstanding_debt  \
37                      15  ...                   5.0           1173.70   
44                      12  ...                   3.0           1095.73   
46                      61  ...                  11.0           2797.17   
60                      54  ...                  12.0           4834.59   
64                      18  ...                   9.0           1037.45   
...                    ...  ...                   ...               ...   
12441                   21  ...                   7.0           1660.14   
12456                   13  ...                   6.0            332.75   
12469                   20  ...                  13.0           4138.67   
12491                   54  ...                   6.0           1453.61   
12492                   27  ...                   2.0            642.46   

       credit_utilization_ratio  credit_history_age  payment_of_min_amount  \
37                    38.170982                28.0                     No   
44                    41.513288                 NaN                     No   
46                    35.201445                14.0                     NM   
60                    30.813295                10.0                    Yes   
64                    32.078619                13.0                    Yes   
...                         ...                 ...                    ...   
12441                 33.883240                16.0                    Yes   
12456                 32.450983                29.0                     No   
12469                 30.807734                 5.0                    Yes   
12491                 34.557510                11.0                    Yes   
12492                 34.026427                29.0                     No   

       total_emi_per_month amount_invested_monthly  \
37              215.839171              105.717873   
44            81050.000000               78.798183   
46               58.868441              133.882578   
60               78.418272              171.639827   
64            50812.000000              231.766467   
...                    ...                     ...   


Se puede asumir que occupation = '_______' representa que no se tiene esta información o no la entrega, o bien, no tiene trabajo estable o formal.

In [509]:
df.loc[:,'payment_of_min_amount'].unique()

array(['No', 'Yes', 'NM'], dtype=object)

In [510]:
mask = df.loc[:,'payment_of_min_amount'] == 'NM'
df.loc[mask,:]

customer_id   age occupation  annual_income  monthly_inhand_salary  \
13     CUS_0x6c66  39.0    Manager       8701.545             519.128750   
35     CUS_0x9b94  36.0    Manager      71681.400            5799.450000   
39     CUS_0x571f  19.0  Developer      66567.320            5800.276667   
46     CUS_0x609d  27.0    _______      14165.230            1057.435833   
47     CUS_0x75c6  30.0    Teacher      87215.680            7489.973333   
...           ...   ...        ...            ...                    ...   
12447  CUS_0x4c83  47.0  Scientist      39362.030            3485.169167   
12448   CUS_0x544  29.0   Mechanic      17013.290            1326.774167   
12458   CUS_0xb09  31.0     Lawyer     146310.680           12124.556667   
12473  CUS_0x62f5  54.0   Musician      99520.500            8479.375000   
12475  CUS_0x1232  17.0  Developer      14937.490            1142.790833   

       num_bank_accounts  num_credit_card  interest_rate  num_of_loan  \
13                     6                5             32          7.0   
35                     3                6             15          1.0   
39                     5                7              5          2.0   
46                     7               10             33          9.0   
47                     2                6              8          2.0   
...                  ...              ...            ...          ...   
12447                  0                3             10          4.0   
12448                813                1             10          1.0   
12458                  5                4              9          1.0   
12473                  3                1           4396          3.0   
12475                  8                8             22          7.0   

       delay_from_due_date  ...  num_credit_inquiries  outstanding_debt  \
13                      23  ...                   9.0           2602.69   
35                       8  ...                   0.0           1293.02   
39                       7  ...                   2.0            233.79   
46                      61  ...                  11.0           2797.17   
47                       2  ...                   0.0            680.77   
...                    ...  ...                   ...               ...   
12447                   13  ...                   8.0           1283.75   
12448                    6  ...                   3.0           1452.79   
12458                    3  ...                   6.0            928.28   
12473                    8  ...                   5.0            547.21   
12475                   43  ...                   9.0           3699.44   

       credit_utilization_ratio  credit_history_age  payment_of_min_amount  \
13                    28.815537                 9.0                     NM   
35                    38.234514                26.0                     NM   
39                    28.453786                 NaN                     NM   
46                    35.201445                14.0                     NM   
47                    35.622883                21.0                     NM   
...                         ...                 ...                    ...   
12447                 27.260894                27.0                     NM   
12448                 28.051684                32.0                     NM   
12458                 31.695064                22.0                     NM   
12473                 38.798114                16.0                     NM   
12475                 23.120963                13.0                     NM   

       total_emi_per_month amount_invested_monthly  \
13               36.548197                     NaN   
35               53.923514              295.811896   
39              100.189361               82.844229   
46               58.868441              133.882578   
47            19734.000000              304.202810   
...                    ...                     ...   


In [511]:
df.loc[:,'payment_behaviour'].unique()

array(['High_spent_Medium_value_payments',
       'Low_spent_Small_value_payments',
       'High_spent_Small_value_payments', '!@9#%8',
       'Low_spent_Medium_value_payments',
       'High_spent_Large_value_payments',
       'Low_spent_Large_value_payments'], dtype=object)

In [512]:
mask = df.loc[:,'payment_behaviour'] == '!@9#%8'
df.loc[mask,:]

customer_id     age    occupation  annual_income  monthly_inhand_salary  \
5      CUS_0x95ee    31.0        Lawyer       73928.46            5988.705000   
22     CUS_0xac86    20.0  Entrepreneur      106733.13            8873.427500   
27     CUS_0x3edc    32.0    Accountant       43070.24                    NaN   
41     CUS_0x6a1b    33.0    Accountant       30788.44            2623.703333   
56     CUS_0x3f5b    25.0        Doctor       80108.31            6866.692500   
...           ...     ...           ...            ...                    ...   
12436  CUS_0x290c  1070.0      Engineer       28954.04            2703.836667   
12462  CUS_0x1ea8    30.0       Teacher       33896.53                    NaN   
12473  CUS_0x62f5    54.0      Musician       99520.50            8479.375000   
12476  CUS_0x64f0    19.0     Architect       39977.21                    NaN   
12499  CUS_0x942c    25.0      Mechanic       39628.99            3359.415833   

       num_bank_accounts  num_credit_card  interest_rate  num_of_loan  \
5                      4                5              8          0.0   
22                     4                4              1          0.0   
27                     3                3             18       -100.0   
41                     7                9             31          6.0   
56                     5                3             17          0.0   
...                  ...              ...            ...          ...   
12436                  8                5             25          5.0   
12462                  7                3             18          2.0   
12473                  3                1           4396          3.0   
12476                  4                7             11          0.0   
12499                  4                6              7          2.0   

       delay_from_due_date  ...  num_credit_inquiries  outstanding_debt  \
5                        8  ...                   NaN            548.20   
22                       7  ...                   1.0             76.23   
27                      11  ...                   8.0           1233.10   
41                      49  ...                   9.0           3470.08   
56                      25  ...                   4.0            997.28   
...                    ...  ...                   ...               ...   
12436                   36  ...                  10.0           2077.68   
12462                   26  ...                   NaN            807.65   
12473                    8  ...                   5.0            547.21   
12476                   13  ...                   4.0            832.09   
12499                   18  ...                   3.0            502.38   

       credit_utilization_ratio  credit_history_age  payment_of_min_amount  \
5                     31.580990                32.0                     No   
22                    39.243640                33.0                     No   
27                    26.365700                19.0                    Yes   
41                    38.254422                 NaN                    Yes   
56                    28.123278                18.0                    Yes   
...                         ...                 ...                    ...   
12436                 37.483994                16.0                    Yes   
12462                 26.800550                27.0                    Yes   
12473                 38.798114                16.0                     NM   
12476                 35.328081                30.0                     No   
12499                 34.192463                31.0                     No   

       total_emi_per_month amount_invested_monthly  payment_behaviour  \
5                 0.000000               42.635590             !@9#%8   
22                0.000000              384.763725             !@9#%8   
27               30.576085              403.516809             !@9#%8   
41              114.533021

Respecto a payment_behaviour = '!@9#%8' no se puede concluir mucho, puede ser que no se tenga información de ello.

## 3 Preparación de los Datos

### 3.1 Preprocesamiento con ColumnTransformer

Convertir columnas mal leídas

In [513]:
df.loc[:,'age'] = pd.to_numeric(df.loc[:,'age'], errors='coerce', downcast='integer')
df.loc[:,'annual_income'] = pd.to_numeric(df.loc[:,'annual_income'], errors='coerce', downcast='integer')
df.loc[:,'num_of_loan'] = pd.to_numeric(df.loc[:,'num_of_loan'], errors='coerce', downcast='integer')
df.loc[:,'delay_from_due_date'] = pd.to_numeric(df.loc[:,'delay_from_due_date'], errors='coerce', downcast='integer')
df.loc[:,'num_of_delayed_payment'] = pd.to_numeric(df.loc[:,'num_of_delayed_payment'], errors='coerce', downcast='integer')
df.loc[:,'changed_credit_limit'] = pd.to_numeric(df.loc[:,'changed_credit_limit'], errors='coerce', downcast='integer')
df.loc[:,'num_credit_inquiries'] = pd.to_numeric(df.loc[:,'num_credit_inquiries'], errors='coerce', downcast='integer')
df.loc[:,'credit_history_age'] = pd.to_numeric(df.loc[:,'credit_history_age'], errors='coerce', downcast='integer')

Generar ColumnTransformer

In [514]:
# Machine Learning - Pipeline
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# Configurar transform_output para que el resultado sea un DataFrame
#from sklearn import set_config
#set_config(transform_output='pandas')

In [515]:
# Machine Learning - Preprocessing
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, OrdinalEncoder
from sklearn.preprocessing import FunctionTransformer

In [516]:
df.dtypes

customer_id                  object
age                         float64
occupation                   object
annual_income               float64
monthly_inhand_salary       float64
num_bank_accounts             int64
num_credit_card               int64
interest_rate                 int64
num_of_loan                 float64
delay_from_due_date           int64
num_of_delayed_payment      float64
changed_credit_limit        float64
num_credit_inquiries        float64
outstanding_debt            float64
credit_utilization_ratio    float64
credit_history_age          float64
payment_of_min_amount        object
total_emi_per_month         float64
amount_invested_monthly     float64
payment_behaviour            object
monthly_balance             float64
credit_score                  int64
dtype: object

In [517]:
log_numeric = [
    'annual_income','delay_from_due_date','outstanding_debt','amount_invested_monthly'
    ]
numeric = [
    'num_bank_accounts','num_of_loan','num_credit_inquiries','total_emi_per_month',
    'credit_utilization_ratio','changed_credit_limit','monthly_balance','num_credit_card',
    'interest_rate','age',
    'annual_income','delay_from_due_date','outstanding_debt','amount_invested_monthly'
    ]
categorical = [
    'occupation', 'payment_of_min_amount', 'payment_behaviour'
    ]
ordinal = [
    ]
not_used = [
    'customer_id','credit_score','monthly_inhand_salary',  'num_of_delayed_payment', 'credit_history_age'
    ]

In [518]:
# Log Transformer
LogTransformer = FunctionTransformer(np.log1p)

Se eligen como transformaciones:
- Codificación One-Hot para variables categóricas.
- Log-transformación para variables con distribución log-normal, descubierta a partir del EDA con y-data profiling.
- Escalamiento robusto para todas las variables numéricas, dada la existencia de valores atípicos o extremos, por lo que se decide hacer tratamiento de ellos mediante RobustScaler.

In [519]:
ct = ColumnTransformer(
    [
        ('One Hot',OneHotEncoder(sparse=False, handle_unknown="ignore",drop='if_binary'),categorical),
        ('Ordinal',OrdinalEncoder(),ordinal),
        ('Log Scaler',LogTransformer,log_numeric),
        ('Scaler',RobustScaler(),numeric)
    ],
    #remainder='passthrough'
).set_output(transform="pandas")

Probar ColumnTransformer:

In [520]:
ct.fit_transform(df)

One Hot__occupation_Accountant  One Hot__occupation_Architect  \
0                                 0.0                            0.0   
1                                 0.0                            0.0   
2                                 0.0                            0.0   
3                                 0.0                            0.0   
4                                 0.0                            0.0   
...                               ...                            ...   
12495                             0.0                            0.0   
12496                             0.0                            0.0   
12497                             0.0                            0.0   
12498                             0.0                            1.0   
12499                             0.0                            0.0   

       One Hot__occupation_Developer  One Hot__occupation_Doctor  \
0                                0.0                         0.0   
1                                0.0                         0.0   
2                                0.0                         0.0   
3                                0.0                         0.0   
4                                1.0                         0.0   
...                              ...                         ...   
12495                            0.0                         0.0   
12496                            0.0                         0.0   
12497                            0.0                         0.0   
12498                            0.0                         0.0   
12499                            0.0                         0.0   

       One Hot__occupation_Engineer  One Hot__occupation_Entrepreneur  \
0                               0.0                               0.0   
1                               0.0                               0.0   
2                               1.0                               0.0   
3                               0.0                               1.0   
4                               0.0                               0.0   
...                             ...                               ...   
12495                           0.0                               0.0   
12496                           0.0                               0.0   
12497                           0.0                               0.0   
12498                           0.0                               0.0   
12499                           0.0                               0.0   

       One Hot__occupation_Journalist  One Hot__occupation_Lawyer  \
0                                 0.0                         0.0   
1                                 0.0                         0.0   
2                                 0.0                         0.0   
3                                 0.0                         0.0   
4                                 0.0                         0.0   
...                               ...                         ...   
12495                             0.0                         1.0   
12496                             0.0                         0.0   
12497                             0.0                         0.0   
12498                             0.0                         0.0   
12499                             0.0                         0.0   

       One Hot__occupation_Manager  One Hot__occupation_Mechanic  ...  \
0                              0.0                           0.0  ...   
1                              0.0                           0.0  ...   
2                              0.0                           0.0  ...   
3                              0.0                           0.0  ...   
4                              0.0                           0.0  ...   
...                            ...                           ...  ...   
12495                          0.0                           0.0  ...   
12496                          0.0 

In [521]:
ct = ColumnTransformer(
    [
        ('One Hot',OneHotEncoder(sparse=False, handle_unknown="ignore",drop='if_binary'),categorical),
        ('Ordinal',OrdinalEncoder(),ordinal),
        ('Log Scaler',LogTransformer,log_numeric),
        ('Scaler',RobustScaler(),numeric)
    ]
)

### 3.2 Holdout

In [522]:
# Train-Test Split
from sklearn.model_selection import train_test_split

In [523]:
# Primero se separan los datos en train y test
features = df.drop(columns=not_used)
y = df.loc[:, 'credit_score']

X_train, X_test, y_train, y_test = train_test_split(
    features, y, test_size=0.2, shuffle=True, stratify=y, random_state=seed
)

### 3.3 Datos Nulos y Tratamiento

Outliers: se tratarán sólo los valores negativos, y en age, se hará un tratamiento especial de los extremos, respecto de las demás, se utilizará un RobustScaler para tratarlos más adelante, ya que en son valores que podrían darse.

Principales limpiezas realizadas:
- age: se asume que los valores negativos son errores de tipeo, por lo que se los convierte a positivos, además, se asume que los valores extremos son errores de tipeo, por lo que se obtiene el residuo de la división por 100, para obtener valores mayores que 0 pero menores a 100.
- variables con valores atípicos negativos: al igual que con la edad, se asume que los valores negativos son errores de tipeo, por lo que se los convierte a positivos, como se mencionó anteriormente, los valores atípicos positivos o extremos, se corregirán con el escalador.
- ocupation: '_______' se asume como NoInfo.
- payment_behaviour: '!@9#%8' también se asume como NoInfo, esto para no perder observaciones.

Cabe mencionar que estas dos últimas, se requiere poder confirmar con mayor conocimiento experto de la generación de las bases de datos, pero como un acercamiento en primera instancia, son supuestos que pueden ser considerados razonables.
- Nulos: se utiliza un SimpleImputer con la media como estrategia.

In [524]:
from sklearn.impute import SimpleImputer
def data_cleaner(X):
    # Outliers: se tratarán sólo los valores negativos, y en age, se hará un tratamiento especial de los extremos, respecto de las demás, se utilizará 
    # un RobustScaler para tratarlos más adelante, ya que en son valores que podrían darse.

    # age: valores negativos y extremos
    # se asume que los valores negativos son errores de tipeo, por lo que se los convierte a positivos
    X.loc[:,'age'] = X.loc[:,'age'].apply(lambda x: -x if x<0 else x)
    # se asume que los valores extremos son errores de tipeo, por lo que se obtiene el residuo de la división por 100
    X.loc[:,'age'] = X.loc[:,'age'].apply(lambda x: x%100 if x>100 else x)
    # se vuelve a valor entero
    X.loc[:,'age'] = pd.to_numeric(X.loc[:,'age'], errors='coerce', downcast='integer')
    
    # negativos
    negative = ['num_bank_accounts','num_of_loan','delay_from_due_date']
    for col in negative:
        X.loc[:,col] = X.loc[:,col].apply(lambda x: -x if x<0 else x)
        X.loc[:,col] = pd.to_numeric(X.loc[:,col], errors='coerce', downcast='integer')

    # occupation = '_______'
    X.loc[:,'occupation'] = X.loc[:,'occupation'].apply(lambda x: 'NoInfo' if x=='_______' else x)

    # payment_behaviour = '!@9#%8' 
    X.loc[:,'payment_behaviour'] = X.loc[:,'payment_behaviour'].apply(lambda x: 'NoInfo' if x=='!@9#%8' else x)


    # Nulos:
    nulls = ['changed_credit_limit','num_credit_inquiries','amount_invested_monthly','monthly_balance']
    imputer = SimpleImputer(strategy='mean')
    X[nulls] = imputer.fit_transform(X[nulls])
    return X

Data Cleaner creado para las variables generadas con Feature Engineering

In [525]:
def data_cleaner_2(X):
    # Outliers: se tratarán sólo los valores negativos, y en age, se hará un tratamiento especial de los extremos, respecto de las demás, se utilizará 
    # un RobustScaler para tratarlos más adelante, ya que en son valores que podrían darse.
    
    # negativos
    negative = ['num_bank_accounts','num_of_loan','delay_from_due_date']
    for col in negative:
        X.loc[:,col] = X.loc[:,col].apply(lambda x: -x if x<0 else x)
        X.loc[:,col] = pd.to_numeric(X.loc[:,col], errors='coerce', downcast='integer')

    # occupation = '_______'
    X.loc[:,'occupation'] = X.loc[:,'occupation'].apply(lambda x: 'NoInfo' if x=='_______' else x)

    # payment_behaviour = '!@9#%8' 
    X.loc[:,'payment_behaviour'] = X.loc[:,'payment_behaviour'].apply(lambda x: 'NoInfo' if x=='!@9#%8' else x)


    # Nulos:
    nulls = ['changed_credit_limit','num_credit_inquiries','amount_invested_monthly','monthly_balance']
    imputer = SimpleImputer(strategy='mean')
    X[nulls] = imputer.fit_transform(X[nulls])
    return X

In [526]:
# Tratamiento de datos
cleaner = FunctionTransformer(data_cleaner)
cleaner_2 = FunctionTransformer(data_cleaner_2)

### 3.4 Feature Engineering (Bonus)

Se crean nuevas variables:

Ratios: 
- debt_to_income_ratio: ratio entre outstanding_debt y annual_income, es decir, cuánto porcentaje del ingreso anual es destinado a pagar deudas impagas.
- credit_card_utilization: ratio ente num_credit_cart y num_bank_accounts, es decir, cuántas tarjetas de crédito hay por cuentas de banco que tiene el cliente.

Discretización de age: en 'young', 'adult', 'middle-aged' y 'senior'.

In [527]:
def Feature_Extractor(X):
    # Crear nuevas características
    X['debt_to_income_ratio'] = X['outstanding_debt'] / X['annual_income']
    X.loc[X['annual_income'] == 0, 'debt_to_income_ratio'] = 0  # Asignar 0 a los casos donde el denominador es 0

    X['credit_card_utilization'] = X['num_credit_card'] / X['num_bank_accounts']
    X.loc[X['num_bank_accounts'] == 0, 'credit_card_utilization'] = 0  # Asignar 0 a los casos donde el denominador es 0

     # age: valores negativos y extremos
    # se asume que los valores negativos son errores de tipeo, por lo que se los convierte a positivos
    X.loc[:,'age'] = X.loc[:,'age'].apply(lambda x: -x if x<0 else x)
    # se asume que los valores extremos son errores de tipeo, por lo que se obtiene el residuo de la división por 100
    X.loc[:,'age'] = X.loc[:,'age'].apply(lambda x: x%100 if x>100 else x)
    # se vuelve a valor entero
    X.loc[:,'age'] = pd.to_numeric(X.loc[:,'age'], errors='coerce', downcast='integer')

    
    # Discretizar la variable 'age'
    X['age_group'] = pd.cut(X['age'], bins=[0, 30, 40, 50, float('inf')],
                            labels=['young', 'adult', 'middle-aged', 'senior'], right=False)
    
    X = X.drop(columns=['age'])
    return X


In [528]:
# Creación del FunctionTransformer
feature_extractor = FunctionTransformer(Feature_Extractor)

Nuevo Column Transformer:
Se agregan las características que se extraen desde el Feature Extractor

In [529]:
# Variables

log_numeric = [
    'annual_income','delay_from_due_date','outstanding_debt','amount_invested_monthly'
    ]
numeric = [
    'num_bank_accounts','num_of_loan','num_credit_inquiries','total_emi_per_month',
    'credit_utilization_ratio','changed_credit_limit','monthly_balance','num_credit_card',
    'interest_rate',
    'annual_income','delay_from_due_date','outstanding_debt','amount_invested_monthly',
    'debt_to_income_ratio', 'credit_card_utilization'
    ]
categorical = [
    'occupation', 'payment_of_min_amount', 'payment_behaviour',
    'age_group'
    ]
ordinal = [
    ]
not_used = [
    'customer_id','credit_score','monthly_inhand_salary',  'num_of_delayed_payment', 'credit_history_age'
    ]

In [530]:
ct = ColumnTransformer(
    [
        ('One Hot',OneHotEncoder(sparse=False, handle_unknown="ignore",drop='if_binary'),categorical),
        ('Ordinal',OrdinalEncoder(),ordinal),
        ('Log Scaler',LogTransformer,log_numeric),
        ('Scaler',RobustScaler(),numeric)
    ],
    #remainder='passthrough'
)

## 4 Baseline

In [531]:
# Machine Learning - Models - Classification
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier

### Pipelines

In [532]:
clf_dummy = Pipeline(steps=[
                    ('extractor', feature_extractor),
                    #('cleaner', cleaner),
                    ('cleaner', cleaner_2),
                    ("preprocessing", ct),
                    ("model", DummyClassifier(strategy='stratified',random_state=seed) )
])

clf_logit = Pipeline(steps=[
                    ('extractor', feature_extractor),
                    #('cleaner', cleaner),
                    ('cleaner', cleaner_2),
                    #('cleaner', cleaner_3),
                    ("preprocessing", ct),
                    ("model", LogisticRegression(random_state=seed) )
])

In [533]:
clf_tree = Pipeline(steps=[
                    ('extractor', feature_extractor),
                    #('cleaner', cleaner),
                    ('cleaner', cleaner_2),
                    ("preprocessing", ct),
                    ("model", DecisionTreeClassifier(random_state=seed) )
])

clf_knn = Pipeline(steps=[
                    ('extractor', feature_extractor),
                    #('cleaner', cleaner),
                    ('cleaner', cleaner_2),
                    ("preprocessing", ct),
                    ("model", KNeighborsClassifier() )
])

clf_svm = Pipeline(steps=[
                    ('extractor', feature_extractor),
                    #('cleaner', cleaner),
                    ('cleaner', cleaner_2),
                    ("preprocessing", ct),
                    ("model", SVC(random_state=seed) )
])

clf_rf = Pipeline(steps=[
                    ('extractor', feature_extractor),
                    #('cleaner', cleaner),
                    ('cleaner', cleaner_2),
                    ("preprocessing", ct),
                    ("model", RandomForestClassifier(random_state=seed) )
])

In [534]:
clf_xgb = Pipeline(steps=[
                    ('extractor', feature_extractor),
                    #('cleaner', cleaner),
                    ('cleaner', cleaner_2),
                    ("preprocessing", ct),
                    ("model", XGBClassifier(random_state=seed) )
])

clf_lgbm = Pipeline(steps=[
                    ('extractor', feature_extractor),
                    #('cleaner', cleaner),
                    ('cleaner', cleaner_2),
                    ("preprocessing", ct),
                    ("model", LGBMClassifier(random_state=seed) )
])

### Evaluación

In [535]:
y.value_counts(normalize=True)

credit_score
0    0.71184
1    0.28816
Name: proportion, dtype: float64

In [536]:
# Machine Learning - Metrics
from sklearn.metrics import f1_score, accuracy_score, precision_score, recall_score, confusion_matrix, classification_report

In [537]:
metricas = pd.DataFrame(columns=['Modelo','Precision_Test'])

In [538]:
# Fit
clf_dummy.fit(X_train,y_train)

# Predict
y_pred = clf_dummy.predict(X_train)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_train))
print(classification_report(y_pred, y_train))
print("\nPrecision-Score", precision_score(y_pred, y_train).round(2))

Matriz de confusión
 [[5118 2103]
 [2000  779]]
              precision    recall  f1-score   support

           0       0.72      0.71      0.71      7221
           1       0.27      0.28      0.28      2779

    accuracy                           0.59     10000
   macro avg       0.49      0.49      0.49     10000
weighted avg       0.59      0.59      0.59     10000


Precision-Score 0.27


In [539]:
# Predict
y_pred = clf_dummy.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(2))

Matriz de confusión
 [[1265  524]
 [ 515  196]]
              precision    recall  f1-score   support

           0       0.71      0.71      0.71      1789
           1       0.27      0.28      0.27       711

    accuracy                           0.58      2500
   macro avg       0.49      0.49      0.49      2500
weighted avg       0.59      0.58      0.59      2500


Precision-Score 0.27


In [540]:
nueva_fila = {'Modelo': 'Dummy', 'Precision_Test': precision_score(y_test, y_pred)} 

# Agregar la nueva fila al DataFrame
metricas.loc[len(metricas)] = nueva_fila


In [541]:
# Fit
clf_logit.fit(X_train,y_train)

# Predict
y_pred = clf_logit.predict(X_train)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_train))
print(classification_report(y_pred, y_train))
print("\nPrecision-Score", precision_score(y_pred, y_train).round(2))

Matriz de confusión
 [[7118 2882]
 [   0    0]]
              precision    recall  f1-score   support

           0       1.00      0.71      0.83     10000
           1       0.00      0.00      0.00         0

    accuracy                           0.71     10000
   macro avg       0.50      0.36      0.42     10000
weighted avg       1.00      0.71      0.83     10000


Precision-Score 0.0


In [542]:
# Predict
y_pred = clf_logit.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(2))

Matriz de confusión
 [[1780  720]
 [   0    0]]
              precision    recall  f1-score   support

           0       1.00      0.71      0.83      2500
           1       0.00      0.00      0.00         0

    accuracy                           0.71      2500
   macro avg       0.50      0.36      0.42      2500
weighted avg       1.00      0.71      0.83      2500


Precision-Score 0.0


In [543]:
nueva_fila = {'Modelo': 'Logit', 'Precision_Test': precision_score(y_pred, y_test)}

# Agregar la nueva fila al DataFrame
metricas.loc[len(metricas)] = nueva_fila

In [544]:
# Fit
clf_tree.fit(X_train,y_train)

# Predict
y_pred = clf_tree.predict(X_train)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_train))
print(classification_report(y_pred, y_train))
print("\nPrecision-Score", precision_score(y_pred, y_train).round(2))

Matriz de confusión
 [[7118    0]
 [   0 2882]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      7118
           1       1.00      1.00      1.00      2882

    accuracy                           1.00     10000
   macro avg       1.00      1.00      1.00     10000
weighted avg       1.00      1.00      1.00     10000


Precision-Score 1.0


In [545]:
# Predict
y_pred = clf_tree.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(2))

Matriz de confusión
 [[1353  374]
 [ 427  346]]
              precision    recall  f1-score   support

           0       0.76      0.78      0.77      1727
           1       0.48      0.45      0.46       773

    accuracy                           0.68      2500
   macro avg       0.62      0.62      0.62      2500
weighted avg       0.67      0.68      0.68      2500


Precision-Score 0.48


In [546]:
nueva_fila = {'Modelo': 'DecisionTree', 'Precision_Test': precision_score(y_pred, y_test)}

# Agregar la nueva fila al DataFrame
metricas.loc[len(metricas)] = nueva_fila

In [547]:
# Fit
clf_knn.fit(X_train,y_train)

# Predict
y_pred = clf_knn.predict(X_train)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_train))
print(classification_report(y_pred, y_train))
print("\nPrecision-Score", precision_score(y_pred, y_train).round(2))

Matriz de confusión
 [[6567 1273]
 [ 551 1609]]
              precision    recall  f1-score   support

           0       0.92      0.84      0.88      7840
           1       0.56      0.74      0.64      2160

    accuracy                           0.82     10000
   macro avg       0.74      0.79      0.76     10000
weighted avg       0.84      0.82      0.83     10000


Precision-Score 0.56


In [548]:
# Predict
y_pred = clf_knn.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(2))

Matriz de confusión
 [[1556  411]
 [ 224  309]]
              precision    recall  f1-score   support

           0       0.87      0.79      0.83      1967
           1       0.43      0.58      0.49       533

    accuracy                           0.75      2500
   macro avg       0.65      0.69      0.66      2500
weighted avg       0.78      0.75      0.76      2500


Precision-Score 0.43


In [549]:
nueva_fila = {'Modelo': 'KNN', 'Precision_Test': precision_score(y_pred, y_test)}

# Agregar la nueva fila al DataFrame
metricas.loc[len(metricas)] = nueva_fila

In [550]:
# Fit
clf_svm.fit(X_train,y_train)

# Predict
y_pred = clf_svm.predict(X_train)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_train))
print(classification_report(y_pred, y_train))
print("\nPrecision-Score", precision_score(y_pred, y_train).round(2))

Matriz de confusión
 [[7118 2881]
 [   0    1]]
              precision    recall  f1-score   support

           0       1.00      0.71      0.83      9999
           1       0.00      1.00      0.00         1

    accuracy                           0.71     10000
   macro avg       0.50      0.86      0.42     10000
weighted avg       1.00      0.71      0.83     10000


Precision-Score 0.0


In [551]:
# Predict
y_pred = clf_svm.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(2))

Matriz de confusión
 [[1780  720]
 [   0    0]]
              precision    recall  f1-score   support

           0       1.00      0.71      0.83      2500
           1       0.00      0.00      0.00         0

    accuracy                           0.71      2500
   macro avg       0.50      0.36      0.42      2500
weighted avg       1.00      0.71      0.83      2500


Precision-Score 0.0


In [552]:
nueva_fila = {'Modelo': 'SVM', 'Precision_Test': precision_score(y_pred, y_test)}

# Agregar la nueva fila al DataFrame
metricas.loc[len(metricas)] = nueva_fila

In [553]:
# Fit
clf_rf.fit(X_train,y_train)

# Predict
y_pred = clf_rf.predict(X_train)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_train))
print(classification_report(y_pred, y_train))
print("\nPrecision-Score", precision_score(y_pred, y_train).round(2))

Matriz de confusión
 [[7118    0]
 [   0 2882]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      7118
           1       1.00      1.00      1.00      2882

    accuracy                           1.00     10000
   macro avg       1.00      1.00      1.00     10000
weighted avg       1.00      1.00      1.00     10000


Precision-Score 1.0


In [554]:
# Predict
y_pred = clf_rf.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(2))

Matriz de confusión
 [[1602  339]
 [ 178  381]]
              precision    recall  f1-score   support

           0       0.90      0.83      0.86      1941
           1       0.53      0.68      0.60       559

    accuracy                           0.79      2500
   macro avg       0.71      0.75      0.73      2500
weighted avg       0.82      0.79      0.80      2500


Precision-Score 0.53


In [555]:
nueva_fila = {'Modelo': 'RandomForest', 'Precision_Test': precision_score(y_pred, y_test)}

# Agregar la nueva fila al DataFrame
metricas.loc[len(metricas)] = nueva_fila

In [556]:
# Fit
clf_xgb.fit(X_train,y_train)

# Predict
y_pred = clf_xgb.predict(X_train)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_train))
print(classification_report(y_pred, y_train))
print("\nPrecision-Score", precision_score(y_pred, y_train).round(2))

Matriz de confusión
 [[6945  459]
 [ 173 2423]]
              precision    recall  f1-score   support

           0       0.98      0.94      0.96      7404
           1       0.84      0.93      0.88      2596

    accuracy                           0.94     10000
   macro avg       0.91      0.94      0.92     10000
weighted avg       0.94      0.94      0.94     10000


Precision-Score 0.84


In [557]:
# Predict
y_pred = clf_xgb.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(2))

Matriz de confusión
 [[1581  346]
 [ 199  374]]
              precision    recall  f1-score   support

           0       0.89      0.82      0.85      1927
           1       0.52      0.65      0.58       573

    accuracy                           0.78      2500
   macro avg       0.70      0.74      0.72      2500
weighted avg       0.80      0.78      0.79      2500


Precision-Score 0.52


In [558]:
nueva_fila = {'Modelo': 'XGBoost', 'Precision_Test': precision_score(y_pred, y_test)}

# Agregar la nueva fila al DataFrame
metricas.loc[len(metricas)] = nueva_fila

In [559]:
# Fit
clf_lgbm.fit(X_train,y_train)

# Predict
y_pred = clf_lgbm.predict(X_train)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_train))
print(classification_report(y_pred, y_train))
print("\nPrecision-Score", precision_score(y_pred, y_train).round(2))

[LightGBM] [Info] Number of positive: 2882, number of negative: 7118
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.002651 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 3593
[LightGBM] [Info] Number of data points in the train set: 10000, number of used features: 49
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.288200 -> initscore=-0.904142
[LightGBM] [Info] Start training from score -0.904142


Matriz de confusión
 [[6684  914]
 [ 434 1968]]
              precision    recall  f1-score   support

           0       0.94      0.88      0.91      7598
           1       0.68      0.82      0.74      2402

    accuracy                           0.87     10000
   macro avg       0.81      0.85      0.83     10000
weighted avg       0.88      0.87      0.87     10000


Precision-Score 0.68


In [560]:
# Predict
y_pred = clf_lgbm.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(2))

Matriz de confusión
 [[1595  323]
 [ 185  397]]
              precision    recall  f1-score   support

           0       0.90      0.83      0.86      1918
           1       0.55      0.68      0.61       582

    accuracy                           0.80      2500
   macro avg       0.72      0.76      0.74      2500
weighted avg       0.82      0.80      0.80      2500


Precision-Score 0.55


In [561]:
nueva_fila = {'Modelo': 'LGBM', 'Precision_Test': precision_score(y_pred, y_test)}

# Agregar la nueva fila al DataFrame
metricas.loc[len(metricas)] = nueva_fila

### Resumen de Modelos

In [575]:
metricas.sort_values(by='Precision_Test', ascending=False)

Modelo  Precision_Test
7          LGBM        0.551389
5  RandomForest        0.529167
6       XGBoost        0.519444
2  DecisionTree        0.480556
3           KNN        0.429167
0         Dummy        0.275668
1         Logit        0.000000
4           SVM        0.000000

- ¿Hay algún clasificador entrenado mejor que el azar (`Dummy`)?
- ¿Cuál es el mejor clasificador entrenado?
- ¿Por qué el mejor clasificador es mejor que los otros?
- Respecto al tiempo de entrenamiento, con cual cree que sería mejor experimentar (piense en el tiempo que le tomaría pasar el modelo por una grilla de optimización de hiperparámetros).

Se aprecia que hay 5 clasificadores mejores que el azar, y es interesante notar que existen 2 que no son capaces de superar este baseline, y no pudiendo capturar ningún cliente mal pagador. El mejor clasificador entrenado es LGBM, logrando una precision en test de 0.55, se puede explicar este resultado debido a que este es un modelo que se entrena sobre muchos clasificadores débiles, los cuales corrigen los errores obtenidos por cada clasificador débil, además de ser un modelo que junto a XGBoost, funcionan bien en datos tabulares, y son de los más sofisticados sin considerar redes neuronales.
Considerando optimizaciones a realizar, y tiempos de entrenamiento, LGBM es un algoritmo liviano, de ahí su nombre, que es trabajado para ser más eficiente en su entrenamiento (0.4 segundos de entrenamiento) que, por ejemplo, un Random Forest (2do lugar, 1.9 segundos en entrenar), haciendo posible entrenar muchos estimadores sin aumentar mucho lso tiempos requeridos.

## 5 Optimización del Modelo

In [576]:
# explicitly require this experimental feature
from sklearn.experimental import enable_halving_search_cv # noqa
# now you can import normally from model_selection
from sklearn.model_selection import HalvingGridSearchCV

Para la grilla de parámetros se configurarán:
- n_estimators: a mayor número de estimadores, en general se producen mejores modelos con mayor poder de generalización.
- max_depth: la profundidad de los árboles que se entrenan, por teoría, en LGBM, se obtienen mejores resultados con muchos estimadores simples.
- learning_rate: parámetro de aprendizaje para la optimización de los modelos.
- ccp_alpha: parámetro de regularización de los árboles.

In [577]:
params = [
    # grid LGMB
    {
        "model": [LGBMClassifier(random_state=seed,n_jobs=-1)],
        'model__n_estimators': [100, 1000, 5000],
        'model__max_depth': [-1, 2, 3],
        'model__learning_rate': [1e-3, 1e-2, 1e-1]
    },
    # grid Random Forest
    {
        "model": [RandomForestClassifier(random_state=seed,n_jobs=-1)],
        'model__n_estimators': [100, 1000, 5000],
        'model__max_depth': [2, 3, 5],
        'model__ccp_alpha': [0.001, 0.1],
    },


]

In [578]:
clf_base = Pipeline(steps=[
                    #('extractor', feature_extractor),
                    #('cleaner', cleaner),
                    ('cleaner', cleaner_2),
                    ("preprocessing", ct),
                    ("model", DummyClassifier(strategy='stratified',random_state=seed) )
])

In [ ]:
search = HalvingGridSearchCV(
    estimator=clf_base,
    param_grid=params,
    verbose=10,
    scoring='precision',
    cv=3,
    error_score='raise',
    n_jobs=-1,
    random_state=seed
)

search.fit(X_train,y_train)

In [580]:
best_model_hgs = search.best_estimator_

In [581]:
best_model_hgs

Pipeline(steps=[('cleaner',
                 FunctionTransformer(func=<function data_cleaner_2 at 0x2aa53a9e0>)),
                ('preprocessing',
                 ColumnTransformer(transformers=[('One Hot',
                                                  OneHotEncoder(drop='if_binary',
                                                                handle_unknown='ignore',
                                                                sparse=False),
                                                  ['occupation',
                                                   'payment_of_min_amount',
                                                   'payment_behaviour',
                                                   'age_group']),
                                                 ('Ordinal', OrdinalEncoder(),
                                                  []),
                                                 ('Log Scaler',
                                                  FunctionTrans...
                                                   'num_credit_inquiries',
                                                   'total_emi_per_month',
                                                   'credit_utilization_ratio',
                                                   'changed_credit_limit',
                                                   'monthly_balance',
                                                   'num_credit_card',
                                                   'interest_rate',
                                                   'annual_income',
                                                   'delay_from_due_date',
                                                   'outstanding_debt',
                                                   'amount_invested_monthly',
                                                   'debt_to_income_ratio',
                                                   'credit_card_utilization'])])),
                ('model',
                 LGBMClassifier(learning_rate=0.01, max_depth=3, n_jobs=-1,
                                random_state=42))])

In [582]:
# Predict
y_pred = best_model_hgs.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(3))

Matriz de confusión
 [[1637  388]
 [ 143  332]]
              precision    recall  f1-score   support

           0       0.92      0.81      0.86      2025
           1       0.46      0.70      0.56       475

    accuracy                           0.79      2500
   macro avg       0.69      0.75      0.71      2500
weighted avg       0.83      0.79      0.80      2500


Precision-Score 0.461


### Bonus: Optuna

In [583]:
!pip install --quiet optuna

In [584]:
import optuna

In [585]:
pipeline = Pipeline(steps=[
    ('extractor', feature_extractor),
    #('cleaner', cleaner),
    ('cleaner', cleaner_2),
    ("preprocessing", ct),
    ("model", LGBMClassifier(random_state=seed, n_jobs=-1))
])

In [586]:
param={
    'model__n_estimators' : optuna.distributions.IntDistribution(100,10000),
    'model__max_depth': optuna.distributions.IntDistribution(1,3),
    'model__learning_rate': optuna.distributions.FloatDistribution(1e-3, 1e1, log=True)
    }

In [587]:
def objective(trial):

    # Entrenar el modelo
    pipeline.fit(X_train, y_train)

    # Calcular la precisión en el conjunto de prueba
    y_pred = pipeline.predict(X_test)

    return precision_score(y_test, y_pred)

# Crear un estudio de Optuna
study = optuna.create_study(direction='maximize')

[I 2023-07-23 00:37:40,718] A new study created in memory with name: no-name-561b3ded-9b54-4800-99ad-a020cbbd7d58


In [ ]:
# Crear un OptunaSearchCV con el Pipeline y el estudio de Optuna
optuna_search = optuna.integration.OptunaSearchCV(
    pipeline,
    param_distributions=param, 
    n_trials=10,  
    scoring='precision',
    cv=3,
    study=study

)

# Realizar la búsqueda de hiperparámetros
optuna_search.fit(X_train, y_train)

# Mejor configuración de hiperparámetros encontrada
best_params = optuna_search.best_params_
best_model = optuna_search.best_estimator_

In [589]:
best_model

Pipeline(steps=[('extractor',
                 FunctionTransformer(func=<function Feature_Extractor at 0x2a97b4e50>)),
                ('cleaner',
                 FunctionTransformer(func=<function data_cleaner_2 at 0x2aa53a9e0>)),
                ('preprocessing',
                 ColumnTransformer(transformers=[('One Hot',
                                                  OneHotEncoder(drop='if_binary',
                                                                handle_unknown='ignore',
                                                                sparse=False),
                                                  ['occupation',
                                                   'payment_of_min_amount',
                                                   'payment_be...
                                                   'credit_utilization_ratio',
                                                   'changed_credit_limit',
                                                   'monthly_balance',
                                                   'num_credit_card',
                                                   'interest_rate',
                                                   'annual_income',
                                                   'delay_from_due_date',
                                                   'outstanding_debt',
                                                   'amount_invested_monthly',
                                                   'debt_to_income_ratio',
                                                   'credit_card_utilization'])])),
                ('model',
                 LGBMClassifier(learning_rate=0.0011740425509363076,
                                max_depth=3, n_estimators=3575, n_jobs=-1,
                                random_state=42))])

In [590]:
# Predict
y_pred = best_model.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(3))

Matriz de confusión
 [[1606  335]
 [ 174  385]]
              precision    recall  f1-score   support

           0       0.90      0.83      0.86      1941
           1       0.53      0.69      0.60       559

    accuracy                           0.80      2500
   macro avg       0.72      0.76      0.73      2500
weighted avg       0.82      0.80      0.80      2500


Precision-Score 0.535


In [591]:
optuna.visualization.plot_contour(study, params=['model__n_estimators','model__max_depth'])

ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed

Figure({
    'data': [{'colorbar': {'title': {'text': 'Objective Value'}},
              'colorscale': [[0.0, 'rgb(247,251,255)'], [0.125,
                             'rgb(222,235,247)'], [0.25, 'rgb(198,219,239)'],
                             [0.375, 'rgb(158,202,225)'], [0.5,
                             'rgb(107,174,214)'], [0.625, 'rgb(66,146,198)'],
                             [0.75, 'rgb(33,113,181)'], [0.875, 'rgb(8,81,156)'],
                             [1.0, 'rgb(8,48,107)']],
              'connectgaps': True,
              'contours': {'coloring': 'heatmap'},
              'hoverinfo': 'none',
              'line': {'smoothing': 1.3},
              'reversescale': False,
              'type': 'contour',
              'x': [0.9, 1, 2, 3, 3.1],
              'y': [987.5, 1394, 1472, 1528, 2878, 3575, 4241, 4734, 7684, 8792,
                    9524, 9930.5],
              'z': [[nan, nan, nan, nan, nan], [nan, nan, 0.6436711125163759, nan,
                    nan], [nan, nan, nan, 0.6578126728064881, nan], [nan, nan, nan,
                    0.631388623687571, nan], [nan, nan, 0.5263007515200059, nan,
                    nan], [nan, nan, nan, 0.6713586624034384, nan], [nan,
                    0.616894173328124, nan, nan, nan], [nan, nan, nan,
                    0.5635271816217537, nan], [nan, nan, nan, 0.3883819582801076,
                    nan], [nan, nan, nan, 0.6637227675131839, nan], [nan, nan, nan,
                    0.5685252063708074, nan], [nan, nan, nan, nan, nan]]},
             {'marker': {'color': 'black', 'line': {'color': 'Grey', 'width': 2.0}},
              'mode': 'markers',
              'showlegend': False,
              'type': 'scatter',
              'x': [2, 3, 1, 3, 3, 3, 3, 3, 2, 3],
              'y': [1394, 1528, 4241, 4734, 1472, 7684, 8792, 3575, 2878, 9524]}],
    'layout': {'template': '...',
               'title': {'text': 'Contour Plot'},
               'xaxis': {'range': [0.9, 3.1], 'title': {'text': 'model__max_depth'}},
               'yaxis': {'range': [987.5, 9930.5], 'title': {'text': 'model__n_estimators'}}}
})

### Imbalanced Learn

Se entrena el mejor modelo de gridsearch y optuna mediante datasets balanceados:

In [592]:
! pip install imbalanced-learn


In [593]:
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import RandomOverSampler
from imblearn.combine import SMOTEENN


In [594]:
undersampler = RandomUnderSampler(random_state=seed)
X_train_resampled, y_train_resampled = undersampler.fit_resample(X_train, y_train)


In [595]:
best_model_hgs.fit(X_train_resampled, y_train_resampled)

# Predict
y_pred = best_model_hgs.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(3))

[LightGBM] [Info] Number of positive: 2882, number of negative: 2882
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000657 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 3349
[LightGBM] [Info] Number of data points in the train set: 5764, number of used features: 49
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000
Matriz de confusión
 [[1402  231]
 [ 378  489]]
              precision    recall  f1-score   support

           0       0.79      0.86      0.82      1633
           1       0.68      0.56      0.62       867

    accuracy                           0.76      2500
   macro avg       0.73      0.71      0.72      2500
weighted avg       0.75      0.76      0.75      2500


Precision-Score 0.679


In [596]:
best_model.fit(X_train_resampled, y_train_resampled)

# Predict
y_pred = best_model.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(3))

[LightGBM] [Info] Number of positive: 2882, number of negative: 2882
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000764 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 3349
[LightGBM] [Info] Number of data points in the train set: 5764, number of used features: 49
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further spli

In [597]:
oversampler = RandomOverSampler(random_state=seed)
X_train_resampled, y_train_resampled = oversampler.fit_resample(X_train, y_train)

In [598]:
best_model_hgs.fit(X_train_resampled, y_train_resampled)

# Predict
y_pred = best_model_hgs.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(3))

[LightGBM] [Info] Number of positive: 7118, number of negative: 7118
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000531 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 3627
[LightGBM] [Info] Number of data points in the train set: 14236, number of used features: 49
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000
Matriz de confusión
 [[1406  232]
 [ 374  488]]
              precision    recall  f1-score   support

           0       0.79      0.86      0.82      1638
           1       0.68      0.57      0.62       862

    accuracy                           0.76      2500
   macro avg       0.73      0.71      0.72      2500
weighted avg       0.75      0.76      0.75      2500


Precision-Score 0.678


In [599]:
best_model.fit(X_train_resampled, y_train_resampled)

# Predict
y_pred = best_model.predict(X_test)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(3))

[LightGBM] [Info] Number of positive: 7118, number of negative: 7118
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000426 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 3627
[LightGBM] [Info] Number of data points in the train set: 14236, number of used features: 49
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with po

Sobre este último mejor modelo, se revisa el threshold, buscando un balance en accuracy y maximizando precision:

In [626]:
# Predict
y_pred = best_model.predict_proba(X_test)[:, 1]

# Definir un nuevo threshold
new_threshold = 0.48

# Realizar predicciones binarias con el nuevo threshold
y_pred = (y_pred >= new_threshold).astype(int)

# Resultados
print("Matriz de confusión\n", confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))
print("\nPrecision-Score", precision_score(y_pred, y_test).round(3))

Matriz de confusión
 [[1370  198]
 [ 410  522]]
              precision    recall  f1-score   support

           0       0.77      0.87      0.82      1568
           1       0.72      0.56      0.63       932

    accuracy                           0.76      2500
   macro avg       0.75      0.72      0.73      2500
weighted avg       0.75      0.76      0.75      2500


Precision-Score 0.725


El mejor treshold probado es 0.48, obteniendo una precision de 0.73

## 6 Interpretabilidad

## 7 Conclusiones